File: scripts/generate-tree-tiles.js
#!/usr/bin/env node
/**
 * Generate tree tile density data from OpenStreetMap PBF files
 *
 * This script processes OSM PBF files and extracts tree data (natural=tree nodes),
 * aggregating counts per z11 tile for use in procedural tree generation.
 *
 * Usage:
 *   node scripts/generate-tree-tiles.js <input.osm.pbf> [output.bin]
 *
 * Output format (binary):
 *   Header (8 bytes):
 *     - Magic: "TREE" (4 bytes)
 *     - Version: uint8 (1 byte)
 *     - Zoom level: uint8 (1 byte)
 *     - Tile count: uint16 LE (2 bytes)
 *
 *   Per tile (7 bytes each):
 *     - X coordinate: uint16 LE (2 bytes)
 *     - Y coordinate: uint16 LE (2 bytes)
 *     - Tree count: uint16 LE (2 bytes) - capped at 65535
 *     - Conifer ratio: uint8 (1 byte) - 0-255 maps to 0.0-1.0
 *
 * Download OSM data from:
 *   - Planet file: https://planet.openstreetmap.org/pbf/planet-latest.osm.pbf
 *   - Regional extracts: https://download.geofabrik.de/
 */

import { createReadStream } from 'fs';
import { writeFile, stat } from 'fs/promises';
import { createGunzip } from 'zlib';
import { pipeline } from 'stream/promises';
import { Transform } from 'stream';

// Configuration
const ZOOM_LEVEL = 11;
const MAX_TILES = 2 ** ZOOM_LEVEL; // 2048 for z11

// Tile data structure
// Map<"x,y" => { total: number, conifers: number }>
const tileData = new Map();

// Stats
let totalTrees = 0;
let processedNodes = 0;

/**
 * Convert lat/lon to tile coordinates at given zoom level
 */
function latLonToTile(lat, lon, zoom) {
  const x = Math.floor((lon + 180) / 360 * (1 << zoom));
  const y = Math.floor((1 - Math.log(Math.tan(lat * Math.PI / 180) + 1 / Math.cos(lat * Math.PI / 180)) / Math.PI) / 2 * (1 << zoom));
  return {
    x: Math.max(0, Math.min(x, (1 << zoom) - 1)),
    y: Math.max(0, Math.min(y, (1 << zoom) - 1))
  };
}

/**
 * Check if a tree is a conifer based on OSM tags
 */
function isConifer(tags) {
  if (!tags) return null; // Unknown

  // Check leaf_type tag
  const leafType = tags.leaf_type || tags['leaf_type'];
  if (leafType === 'needleleaved') return true;
  if (leafType === 'broadleaved') return false;

  // Check genus for common conifers
  const genus = (tags.genus || tags['genus'] || '').toLowerCase();
  const coniferGenera = [
    'pinus', 'picea', 'abies', 'larix', 'cedrus', 'juniperus',
    'thuja', 'cupressus', 'sequoia', 'taxus', 'tsuga', 'pseudotsuga',
    'cryptomeria', 'araucaria', 'metasequoia', 'chamaecyparis'
  ];
  if (coniferGenera.some(g => genus.includes(g))) return true;

  // Check species for common conifers
  const species = (tags.species || tags['species'] || tags.taxon || '').toLowerCase();
  if (coniferGenera.some(g => species.includes(g))) return true;

  return null; // Unknown
}

/**
 * Process a single OSM node
 */
function processNode(node) {
  processedNodes++;

  if (processedNodes % 10000000 === 0) {
    console.log(`  Processed ${(processedNodes / 1000000).toFixed(1)}M nodes, found ${totalTrees} trees in ${tileData.size} tiles...`);
  }

  // Check if this is a tree
  if (!node.tags || node.tags.natural !== 'tree') {
    return;
  }

  if (node.lat === undefined || node.lon === undefined) {
    return;
  }

  totalTrees++;

  // Get tile coordinates
  const tile = latLonToTile(node.lat, node.lon, ZOOM_LEVEL);
  const key = `${tile.x},${tile.y}`;

  // Update tile data
  if (!tileData.has(key)) {
    tileData.set(key, { total: 0, conifers: 0, deciduous: 0 });
  }

  const data = tileData.get(key);
  data.total++;

  const conifer = isConifer(node.tags);
  if (conifer === true) {
    data.conifers++;
  } else if (conifer === false) {
    data.deciduous++;
  }
}

/**
 * Parse OSM PBF format
 * PBF format: https://wiki.openstreetmap.org/wiki/PBF_Format
 */
async function parsePBF(inputPath) {
  // Dynamic import for optional dependency
  let osmpbf;
  try {
    osmpbf = await import('osm-pbf-parser');
  } catch (e) {
    console.error('Error: osm-pbf-parser not installed.');
    console.error('Run: npm install osm-pbf-parser');
    process.exit(1);
  }

  const through = (await import('through2')).default;

  console.log(`Parsing PBF file: ${inputPath}`);

  const fileStats = await stat(inputPath);
  console.log(`File size: ${(fileStats.size / 1024 / 1024 / 1024).toFixed(2)} GB`);

  return new Promise((resolve, reject) => {
    const parser = osmpbf.default ? osmpbf.default() : osmpbf();

    createReadStream(inputPath)
      .pipe(parser)
      .pipe(through.obj(function(items, enc, next) {
        for (const item of items) {
          if (item.type === 'node') {
            processNode(item);
          }
        }
        next();
      }))
      .on('finish', resolve)
      .on('error', reject);
  });
}

/**
 * Parse OSM XML format (for smaller files or testing)
 */
async function parseXML(inputPath) {
  const sax = await import('sax');
  const fs = await import('fs');

  console.log(`Parsing XML file: ${inputPath}`);

  return new Promise((resolve, reject) => {
    const parser = sax.createStream(true);
    let currentNode = null;

    parser.on('opentag', (tag) => {
      if (tag.name === 'node') {
        currentNode = {
          lat: parseFloat(tag.attributes.lat),
          lon: parseFloat(tag.attributes.lon),
          tags: {}
        };
      } else if (tag.name === 'tag' && currentNode) {
        currentNode.tags[tag.attributes.k] = tag.attributes.v;
      }
    });

    parser.on('closetag', (name) => {
      if (name === 'node' && currentNode) {
        processNode(currentNode);
        currentNode = null;
      }
    });

    parser.on('end', resolve);
    parser.on('error', reject);

    let stream = fs.createReadStream(inputPath);
    if (inputPath.endsWith('.gz')) {
      stream = stream.pipe(createGunzip());
    }
    stream.pipe(parser);
  });
}

/**
 * Write binary output file
 */
async function writeBinaryOutput(outputPath) {
  console.log(`\nWriting binary output to: ${outputPath}`);

  // Sort tiles by morton code for better compression potential
  const tiles = Array.from(tileData.entries()).map(([key, data]) => {
    const [x, y] = key.split(',').map(Number);
    // Calculate conifer ratio - if we have known types, use those; otherwise default to 0.3
    let coniferRatio;
    const known = data.conifers + data.deciduous;
    if (known > 0) {
      coniferRatio = data.conifers / known;
    } else {
      coniferRatio = 0.3; // Default assumption
    }
    return { x, y, count: data.total, coniferRatio };
  });

  // Sort by y then x for locality
  tiles.sort((a, b) => a.y !== b.y ? a.y - b.y : a.x - b.x);

  // Calculate buffer size
  const headerSize = 8;
  const tileSize = 7;
  const bufferSize = headerSize + tiles.length * tileSize;

  const buffer = Buffer.alloc(bufferSize);
  let offset = 0;

  // Write header
  buffer.write('TREE', offset); offset += 4;           // Magic
  buffer.writeUInt8(1, offset); offset += 1;           // Version
  buffer.writeUInt8(ZOOM_LEVEL, offset); offset += 1;  // Zoom level
  buffer.writeUInt16LE(tiles.length, offset); offset += 2; // Tile count (might overflow for planet!)

  // For planet-scale data, we might have more than 65535 tiles
  // Let's use a 32-bit count instead
  // Rewrite with 32-bit count
  offset = 0;
  const bufferSizeV2 = 10 + tiles.length * tileSize;
  const bufferV2 = Buffer.alloc(bufferSizeV2);

  bufferV2.write('TREE', offset); offset += 4;           // Magic
  bufferV2.writeUInt8(2, offset); offset += 1;           // Version 2
  bufferV2.writeUInt8(ZOOM_LEVEL, offset); offset += 1;  // Zoom level
  bufferV2.writeUInt32LE(tiles.length, offset); offset += 4; // Tile count (32-bit)

  // Write tiles
  for (const tile of tiles) {
    bufferV2.writeUInt16LE(tile.x, offset); offset += 2;
    bufferV2.writeUInt16LE(tile.y, offset); offset += 2;
    bufferV2.writeUInt16LE(Math.min(tile.count, 65535), offset); offset += 2;
    bufferV2.writeUInt8(Math.round(tile.coniferRatio * 255), offset); offset += 1;
  }

  await writeFile(outputPath, bufferV2);

  console.log(`\nOutput statistics:`);
  console.log(`  Total trees: ${totalTrees.toLocaleString()}`);
  console.log(`  Tiles with trees: ${tiles.length.toLocaleString()}`);
  console.log(`  File size: ${(bufferV2.length / 1024).toFixed(2)} KB`);
  console.log(`  Average trees per tile: ${(totalTrees / tiles.length).toFixed(1)}`);

  // Print top 10 densest tiles
  const densest = [...tiles].sort((a, b) => b.count - a.count).slice(0, 10);
  console.log(`\n  Top 10 densest tiles:`);
  for (const t of densest) {
    console.log(`    z${ZOOM_LEVEL}/${t.x}/${t.y}: ${t.count} trees (${(t.coniferRatio * 100).toFixed(0)}% conifer)`);
  }
}

/**
 * Main entry point
 */
async function main() {
  const args = process.argv.slice(2);

  if (args.length < 1) {
    console.log(`
Usage: node scripts/generate-tree-tiles.js <input> [output]

Arguments:
  input   - OSM data file (.osm.pbf, .osm, .osm.gz)
  output  - Output binary file (default: public/tree-tiles.bin)

Examples:
  # Process a regional extract
  node scripts/generate-tree-tiles.js germany-latest.osm.pbf

  # Process planet file
  node scripts/generate-tree-tiles.js planet-latest.osm.pbf public/tree-tiles.bin

Download OSM data from:
  - Planet: https://planet.openstreetmap.org/pbf/
  - Regions: https://download.geofabrik.de/
`);
    process.exit(1);
  }

  const inputPath = args[0];
  const outputPath = args[1] || 'public/tree-tiles.bin';

  console.log('='.repeat(60));
  console.log('OSM Tree Tile Generator');
  console.log('='.repeat(60));
  console.log(`Input: ${inputPath}`);
  console.log(`Output: ${outputPath}`);
  console.log(`Zoom level: ${ZOOM_LEVEL}`);
  console.log('');

  const startTime = Date.now();

  // Parse based on file extension
  if (inputPath.endsWith('.pbf')) {
    await parsePBF(inputPath);
  } else if (inputPath.endsWith('.osm') || inputPath.endsWith('.osm.gz')) {
    await parseXML(inputPath);
  } else {
    console.error(`Unsupported file format: ${inputPath}`);
    console.error('Supported formats: .osm.pbf, .osm, .osm.gz');
    process.exit(1);
  }

  // Write output
  await writeBinaryOutput(outputPath);

  const elapsed = (Date.now() - startTime) / 1000;
  console.log(`\nCompleted in ${elapsed.toFixed(1)} seconds`);
}

main().catch(err => {
  console.error('Error:', err);
  process.exit(1);
});

File: scripts/generate-tree-tiles.py
#!/usr/bin/env python3
# /// script
# requires-python = ">=3.9"
# dependencies = [
#     "osmium>=4.0.0",
# ]
# ///
"""
Generate tree tile density data from OpenStreetMap PBF files.

This script processes OSM PBF files using pyosmium (fast C++ bindings)
and generates a tree-tiles.bin file with tree density per z11 tile.

Usage:
    uv run scripts/generate-tree-tiles.py <input.osm.pbf> [output.bin]

    # Example with planet file:
    uv run scripts/generate-tree-tiles.py ~/Downloads/planet-latest.osm.pbf

Download OSM data from:
    - Planet: https://planet.openstreetmap.org/pbf/planet-latest.osm.pbf
    - Regions: https://download.geofabrik.de/
"""

import math
import struct
import sys
import time
from pathlib import Path

import osmium

# Unbuffered output for progress visibility
sys.stdout.reconfigure(line_buffering=True)

# Configuration
ZOOM_LEVEL = 14
DEFAULT_OUTPUT = Path(__file__).parent.parent / "public" / "tree-tiles.bin"

# Conifer genera for classification (lowercase for comparison)
CONIFER_GENERA = {
    "pinus", "picea", "abies", "larix", "cedrus", "juniperus",
    "thuja", "cupressus", "sequoia", "taxus", "tsuga", "pseudotsuga",
    "cryptomeria", "araucaria", "metasequoia", "chamaecyparis"
}


def lat_lon_to_tile(lat: float, lon: float, zoom: int) -> tuple[int, int]:
    """Convert lat/lon to tile coordinates at given zoom level."""
    # Clamp latitude to valid Web Mercator range
    lat = max(-85.051129, min(85.051129, lat))
    n = 1 << zoom
    x = int((lon + 180.0) / 360.0 * n)
    y = int((1.0 - math.log(math.tan(math.radians(lat)) + 1.0 / math.cos(math.radians(lat))) / math.pi) / 2.0 * n)
    return (max(0, min(x, n - 1)), max(0, min(y, n - 1)))


def is_conifer(tags) -> bool | None:
    """Check if a tree is a conifer based on OSM tags. Returns None if unknown."""
    leaf_type = tags.get("leaf_type", "")
    if leaf_type == "needleleaved":
        return True
    if leaf_type == "broadleaved":
        return False

    genus = tags.get("genus", "").lower()
    if genus and any(g in genus for g in CONIFER_GENERA):
        return True

    species = tags.get("species", tags.get("taxon", "")).lower()
    if species and any(g in species for g in CONIFER_GENERA):
        return True

    return None


class TreeHandler(osmium.SimpleHandler):
    """Handler that processes tree nodes and accumulates tile counts."""

    def __init__(self):
        super().__init__()
        self.tile_counts: dict[tuple[int, int], dict] = {}
        self.total_trees = 0
        self.processed_nodes = 0
        self.start_time = time.time()
        self.last_report = 0

    def node(self, n):
        self.processed_nodes += 1

        # Progress report every 10M nodes
        if self.processed_nodes - self.last_report >= 10_000_000:
            self.last_report = self.processed_nodes
            elapsed = time.time() - self.start_time
            print(f"  Processed {self.processed_nodes // 1_000_000}M nodes, "
                  f"found {self.total_trees:,} trees in {len(self.tile_counts):,} tiles ({elapsed:.0f}s)")

        # Check if this is a tree
        if n.tags.get("natural") != "tree":
            return

        if not n.location.valid():
            return

        self.total_trees += 1

        # Get tile coordinates
        tile = lat_lon_to_tile(n.location.lat, n.location.lon, ZOOM_LEVEL)

        # Update tile data
        if tile not in self.tile_counts:
            self.tile_counts[tile] = {"total": 0, "conifers": 0, "deciduous": 0}

        data = self.tile_counts[tile]
        data["total"] += 1

        conifer = is_conifer(n.tags)
        if conifer is True:
            data["conifers"] += 1
        elif conifer is False:
            data["deciduous"] += 1


def write_binary_output(output_path: Path, tile_counts: dict) -> None:
    """Write binary output file in the tree-tiles.bin format."""
    print(f"\nWriting binary output to: {output_path}")

    # Convert to list and calculate conifer ratios
    tiles = []
    for (tile_x, tile_y), data in tile_counts.items():
        known = data["conifers"] + data["deciduous"]
        conifer_ratio = data["conifers"] / known if known > 0 else 0.3
        tiles.append({
            "tile_x": tile_x,
            "tile_y": tile_y,
            "tree_count": data["total"],
            "conifer_ratio": conifer_ratio,
        })

    # Sort by y then x for locality
    tiles.sort(key=lambda t: (t["tile_y"], t["tile_x"]))

    # Binary format:
    # Header (10 bytes): "TREE" + version (1) + zoom (1) + tile_count (4)
    # Per tile (7 bytes): x (2) + y (2) + count (2) + conifer_ratio (1)
    header_size = 10
    tile_size = 7
    buffer = bytearray(header_size + len(tiles) * tile_size)

    # Write header
    offset = 0
    buffer[0:4] = b"TREE"
    offset += 4
    buffer[offset] = 2  # Version
    offset += 1
    buffer[offset] = ZOOM_LEVEL
    offset += 1
    struct.pack_into("<I", buffer, offset, len(tiles))
    offset += 4

    # Write tiles
    total_trees = 0
    for tile in tiles:
        struct.pack_into("<H", buffer, offset, tile["tile_x"])
        offset += 2
        struct.pack_into("<H", buffer, offset, tile["tile_y"])
        offset += 2
        count = min(tile["tree_count"], 65535)
        struct.pack_into("<H", buffer, offset, count)
        offset += 2
        buffer[offset] = round(tile["conifer_ratio"] * 255)
        offset += 1
        total_trees += tile["tree_count"]

    output_path.write_bytes(buffer)

    print(f"\nOutput statistics:")
    print(f"  Total trees: {total_trees:,}")
    print(f"  Tiles with trees: {len(tiles):,}")
    print(f"  File size: {len(buffer) / 1024:.2f} KB")
    if tiles:
        print(f"  Average trees per tile: {total_trees / len(tiles):.1f}")


def main():
    if len(sys.argv) < 2:
        print(__doc__)
        print("Error: Input PBF file required")
        print("\nExample:")
        print("  uv run scripts/generate-tree-tiles.py ~/Downloads/planet-latest.osm.pbf")
        sys.exit(1)

    input_path = Path(sys.argv[1])
    output_path = Path(sys.argv[2]) if len(sys.argv) > 2 else DEFAULT_OUTPUT

    if not input_path.exists():
        print(f"Error: Input file not found: {input_path}")
        sys.exit(1)

    print("=" * 60)
    print("OSM Tree Tile Generator (pyosmium)")
    print("=" * 60)
    print(f"Input: {input_path}")
    print(f"Output: {output_path}")
    print(f"Zoom level: {ZOOM_LEVEL}")
    file_size_gb = input_path.stat().st_size / (1024**3)
    print(f"File size: {file_size_gb:.2f} GB")
    print()

    start_time = time.time()

    # Process PBF file
    print("Processing PBF file...")
    handler = TreeHandler()
    handler.apply_file(str(input_path), locations=True)

    elapsed = time.time() - start_time
    print(f"\nProcessing completed in {elapsed:.1f} seconds")
    print(f"Processed {handler.processed_nodes:,} nodes")
    print(f"Found {handler.total_trees:,} trees in {len(handler.tile_counts):,} tiles")

    # Write output
    write_binary_output(output_path, handler.tile_counts)

    total_elapsed = time.time() - start_time
    print(f"\nCompleted in {total_elapsed:.1f} seconds")


if __name__ == "__main__":
    main()
